eventorder_test.go

```diff
@@ -0,0 +1,403 @@
+		transformerContainer, err = destination.SetupTransformer(pool, t)
+		return err
+	})
+	require.NoError(t, containersGroup.Wait())
+	t.Logf("Started docker services")
+
+	t.Logf("Starting webhook destination server")
+	webhook := m.newWebhook(t, spec)
+	defer webhook.Server.Close()
+
+	t.Logf("Preparing rudder-server config")
+	writeKey := trand.String(27)
+	workspaceID := trand.String(27)
+	templateCtx := map[string]string{
+		"webhookUrl":  webhook.Server.URL,
+		"writeKey":    writeKey,
+		"workspaceId": workspaceID,
+	}
+	configJsonPath := workspaceConfig.CreateTempFile(t, "testdata/eventOrderTestTemplate.json", templateCtx)
+	t.Setenv("RSERVER_BACKEND_CONFIG_CONFIG_FROM_FILE", "true")
+	t.Setenv("RSERVER_BACKEND_CONFIG_CONFIG_JSONPATH", configJsonPath)
+
+	t.Logf("Starting rudder-server")
+	t.Setenv("DEPLOYMENT_TYPE", string(deployment.DedicatedType))
+
+	t.Setenv("JOBS_DB_PORT", postgresContainer.Port)
+	t.Setenv("JOBS_DB_USER", postgresContainer.User)
+	t.Setenv("JOBS_DB_DB_NAME", postgresContainer.Database)
+	t.Setenv("JOBS_DB_PASSWORD", postgresContainer.Password)
+	t.Setenv("DEST_TRANSFORM_URL", transformerContainer.TransformURL)
+
+	t.Setenv("RSERVER_WAREHOUSE_MODE", "off")
+	t.Setenv("RSERVER_ENABLE_STATS", "false")
+	t.Setenv("RSERVER_JOBS_DB_BACKUP_ENABLED", "false")
+
+	// generatorLoop
+	t.Setenv("RSERVER_ROUTER_JOB_QUERY_BATCH_SIZE", "500")
+	t.Setenv("RSERVER_ROUTER_READ_SLEEP", "10ms")
+	t.Setenv("RSERVER_ROUTER_MIN_RETRY_BACKOFF", "5ms")
+	t.Setenv("RSERVER_ROUTER_MAX_RETRY_BACKOFF", "10ms")
+	t.Setenv("RSERVER_ROUTER_ALLOW_ABORTED_USER_JOBS_COUNT_FOR_PROCESSING", "1")
+
+	// worker
+	t.Setenv("RSERVER_ROUTER_NO_OF_JOBS_PER_CHANNEL", "100")
+	t.Setenv("RSERVER_ROUTER_JOBS_BATCH_TIMEOUT", "100ms")
+
+	// statusInsertLoop
+	t.Setenv("RSERVER_ROUTER_UPDATE_STATUS_BATCH_SIZE", "100")
+	t.Setenv("RSERVER_ROUTER_MAX_STATUS_UPDATE_WAIT", "10ms")
+
+	// hack
+	// t.Setenv("RSERVER_ROUTER_GUARANTEE_USER_EVENT_ORDER", "false")
+
+	// find free ports for gateway and admin http server to listen on
+	httpPortInt, err := freeport.GetFreePort()
+	require.NoError(t, err)
+	gatewayPort = strconv.Itoa(httpPortInt)
+	// httpAdminPort, err := freeport.GetFreePort()
+	// require.NoError(t, err)
+
+	t.Setenv("RSERVER_GATEWAY_WEB_PORT", gatewayPort)
+	t.Setenv("RUDDER_TMPDIR", os.TempDir())
+
+	svcDone := make(chan struct{})
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				if passed {
+					t.Logf("server panicked: %v", r)
+					close(svcDone)
+				} else {
+					t.Errorf("server panicked: %v", r)
+				}
+			}
+		}()
+		_ = main.Run(ctx)
+		close(svcDone)
+	}()
+
+	t.Logf("waiting rudder-server to start properly")
+	require.Eventually(t, func() bool {
+		url := fmt.Sprintf("http://localhost:%s/health", gatewayPort)
+		if resp, err := http.Get(url); err == nil {
+			resp.Body.Close()
+			return resp.StatusCode == 200
+		}
+		return false
+	}, 20*time.Second, 100*time.Millisecond, "server should be able to start")
```

Comment:
[question] do you feel it makes sense to standardise the timeouts and polling periods by adding a central value for such timeouts and polling periods? this can make us wait for the same amount of time for server to start in different tests.

Reply:
if we standardize, this timeout needs to be adequately large to cover the worst case scenario. Please note that not all server configurations will require the same amount of time to startup.
What we do need is a test abstraction to start a rudder-server, like what we have now for all the rest (rudder-transformer, destinations, etc.). But for this to happen we need to move `Run` method outside the `main` package.